# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

### Q1
 Identify the dimensions of the dataset by determining the number of rows and columns it contains.


In [100]:
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
import numpy as np
import pandas as pd
df = pd.read_csv(url)
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [101]:
df = df.dropna(how='all')
df = df.reset_index(drop=True)
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

### Q2
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.


In [102]:
# Identify numerical and categorical columns
# Professional tip: Always check data types to avoid calculation errors later
print(df.info())

# Check for null values
null_counts = df.isnull().sum()
print(f"Missing values per column:\n{null_counts}")
print(f"Audit : Nombre de lignes après nettoyage : {len(df)}") # Doit afficher ~1071

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 92.2+ KB
None
Missing values per column:
Customer                       0
ST                             0
GENDER        

In [103]:
df.rename(columns={'GENDER': 'Gender'}, inplace=True)
df.rename(columns={'ST': 'State'}, inplace=True)
#*In education, "Bachelors" could be replaced by "Bachelor"*
df.loc[df["Education"] == "Bachelors", "Education"] = "Bachelor"
#Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
df.loc[df["Gender"] == "F", "Gender"] = "Female"
df.loc[df["Gender"] == "Femal", "Gender"] = "Female"
df.loc[df["Gender"] == "female", "Gender"] = "Female"
df.loc[df["Gender"] == "M", "Gender"] = "Male"
#State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
df.loc[df["State"] == "AZ", "State"] = "Arizona"
df.loc[df["State"] == "Cali", "State"] = "California"
df.loc[df["State"] == "WA", "State"] = "Washington"
df['State'] = df['State'].fillna("Unknown")
#In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*
df.loc[df["Vehicle Class"] == "Sports Car", "Vehicle Class"] = "Luxury"
df.loc[df["Vehicle Class"] == "Luxury SUV", "Vehicle Class"] = "Luxury"
df.loc[df["Vehicle Class"] == "Luxury Car", "Vehicle Class"] = "Luxury"
df.loc[df["Vehicle Class"] == "SUV", "Vehicle Class"] = "Luxury"

df.columns = [col.lower().replace(" ", "_") for col in df.columns]

list(df.columns)

['customer',
 'state',
 'gender',
 'education',
 'customer_lifetime_value',
 'income',
 'monthly_premium_auto',
 'number_of_open_complaints',
 'policy_type',
 'vehicle_class',
 'total_claim_amount']

In [104]:
df.head()

customer       state  gender             education customer_lifetime_value  \
0  RB50392  Washington     NaN                Master                     NaN   
1  QZ44356     Arizona  Female              Bachelor              697953.59%   
2  AI49188      Nevada  Female              Bachelor             1288743.17%   
3  WW63253  California    Male              Bachelor              764586.18%   
4  GA49547  Washington    Male  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3         Luxury          529.881344  
4  Four-Door Car           17.269323

In [105]:
df.loc[:, 'customer_lifetime_value'] = (
    df['customer_lifetime_value']
    .astype(str)                        # force the texte
    .str.replace('%', '', regex=False)  # supprenss %
    .astype(float)        # back to float
)

df.loc[:, 'total_claim_amount'] = (
    df['total_claim_amount']
    .astype(str)                        # force the texte
    .astype(float)                     # back to float
)
df.loc[:, 'monthly_premium_auto'] = (
    df['monthly_premium_auto']
    .astype(str)                        # force the texte
    .str.replace(',', '.').str.strip()
)
df['monthly_premium_auto'] = pd.to_numeric(df['monthly_premium_auto'], errors='coerce')  # back to float

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     954 non-null    object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   object 
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 92.2+ KB
None


C:\Users\Utilisateur\AppData\Local\Temp\ipykernel_21948\1462203819.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['1000.0' '94.0' '108.0' ... '88.0' '114.0' '112.0']' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc[:, 'monthly_premium_auto'] = (


In [106]:
df['state'] = df['state'].fillna("Unknown")
df['state'] = df['state'].fillna("Undefined")
df.head()

customer       state  gender             education customer_lifetime_value  \
0  RB50392  Washington     NaN                Master                     NaN   
1  QZ44356     Arizona  Female              Bachelor               697953.59   
2  AI49188      Nevada  Female              Bachelor              1288743.17   
3  WW63253  California    Male              Bachelor               764586.18   
4  GA49547  Washington    Male  High School or Below               536307.65   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3         Luxury          529.881344  
4  Four-Door Car           17.269323

In [107]:
# 1. On retire les espaces et on remplace la virgule par un point
df['monthly_premium_auto'] = df['monthly_premium_auto'].astype(str).str.replace(',', '.').str.strip()

# 2. On convertit en numérique (les erreurs deviennent des NaN)
df['monthly_premium_auto'] = pd.to_numeric(df['monthly_premium_auto'], errors='coerce')

# 3. On vérifie le type
print(f"Type de monthly_premium_auto: {df['monthly_premium_auto'].dtype}")

Type de monthly_premium_auto: float64


### Q3 
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.


In [108]:
print(f"Number of unique states: {df.state.nunique()}")
print(f"Number of unique customers: {df.customer.nunique()}")
print(f"Number of unique education levels: {df.education.nunique()}")
print(f"Number of unique genders: {df.gender.nunique()}")
print(f"Number of unique customer lifetime values: {df.customer_lifetime_value.nunique()}")
print(f"Number of unique incomes: {df.income.nunique()}")
print(f"Number of unique number of open complaints: {df.number_of_open_complaints.nunique()}")
print(f"Number of unique policy types: {df.policy_type.nunique()}")
print(f"Number of unique vehicle classes: {df.vehicle_class.nunique()}")
print(f"Number of unique total claim amounts: {df.total_claim_amount.nunique()}")
print(f"Number of unique monthly premium autos: {df.monthly_premium_auto.nunique()}")

Number of unique states: 5
Number of unique customers: 1071
Number of unique education levels: 5
Number of unique genders: 2
Number of unique customer lifetime values: 1027
Number of unique incomes: 774
Number of unique number of open complaints: 6
Number of unique policy types: 3
Number of unique vehicle classes: 3
Number of unique total claim amounts: 761
Number of unique monthly premium autos: 132


In [109]:
#['customer', 'state',  gender', 'education'','policy_type','vehicle_class',''] those are categorical columns
print(f"mode(s) of state: {df['state'].mode().values}")
print(f"mode(s) of gender: {df['gender'].mode().values}")
print(f"mode(s) of education: {df['education'].mode().values}")
print(f"mode(s) of policy_type: {df['policy_type'].mode().values}")
print(f"mode(s) of vehicle_class: {df['vehicle_class'].mode().values}")
print(f'the representative customer is a {df['gender'].mode().values} graduate with a {df['education'].mode().values}, living in {df['state'].mode().values}, drinving her {df['policy_type'].mode().values}, like {df['vehicle_class'].mode().values}')


mode(s) of state: ['California']
mode(s) of gender: ['Female']
mode(s) of education: ['Bachelor']
mode(s) of policy_type: ['Personal Auto']
mode(s) of vehicle_class: ['Four-Door Car']
the representative customer is a ['Female'] graduate with a ['Bachelor'], living in ['California'], drinving her ['Personal Auto'], like ['Four-Door Car']


In [110]:
#['customer_lifetime_value','income','monthly_premium_auto','number_of_open_complaints,'total_claim_amount'] those are numerical columns
print(df['customer_lifetime_value'].describe())
print(df['income'].describe())


count       1068.0
unique      1027.0
top       251459.2
freq           4.0
Name: customer_lifetime_value, dtype: float64
count     1071.000000
mean     39295.701214
std      30469.427060
min          0.000000
25%      14072.000000
50%      36234.000000
75%      64631.000000
max      99960.000000
Name: income, dtype: float64


In [111]:
#['customer_lifetime_value','income','monthly_premium_auto','number_of_open_complaints,'total_claim_amount'] those are numerical columns
print(df['monthly_premium_auto'].describe())
print(df['total_claim_amount'].describe())

count     1071.000000
mean       193.234360
std       1601.190369
min         61.000000
25%         68.000000
50%         83.000000
75%        109.500000
max      35354.000000
Name: monthly_premium_auto, dtype: float64
count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: total_claim_amount, dtype: float64


In [112]:
print(f"mode(s) of customer_lifetime_value: {df['customer_lifetime_value'].mode().values}")
print(f"mode(s) of income: {df['income'].mode().values}")
print(f"mode(s) of monthly_premium_auto: {df['monthly_premium_auto'].mode().values}")
print(f"mode(s) of total_claim_amount: {df['total_claim_amount'].mode().values}")
print(f"mode(s) of number_of_open_complaints: {df['number_of_open_complaints'].mode().values}")

mode(s) of customer_lifetime_value: [251459.2 445811.34]
mode(s) of income: [0.]
mode(s) of monthly_premium_auto: [65.]
mode(s) of total_claim_amount: [321.6]
mode(s) of number_of_open_complaints: ['1/0/00']


### Q4
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.


#### the central tendency : mean, median, mode,

In [113]:
# the central tendency
print(f"mode(s) of customer_lifetime_value: {df['customer_lifetime_value'].mode().values}")
print(f"mode(s) of income: {df['income'].mode().values}")
print(f"mode(s) of monthly_premium_auto: {df['monthly_premium_auto'].mode().values}")
print(f"mode(s) of total_claim_amount: {df['total_claim_amount'].mode().values}")

print(f"median of customer_lifetime_value: {df['customer_lifetime_value'].median()}")
print(f"median of income: {df['income'].median()}")
print(f"median of monthly_premium_auto: {df['monthly_premium_auto'].median()}")
print(f"median of total_claim_amount: {df['total_claim_amount'].median()}")

print(f"mean of customer_lifetime_value: {df['customer_lifetime_value'].mean()}")
print(f"mean of income: {df['income'].mean()}")
print(f"mean of monthly_premium_auto: {df['monthly_premium_auto'].mean()}")
print(f"mean of total_claim_amount: {df['total_claim_amount'].mean()}")

mode(s) of customer_lifetime_value: [251459.2 445811.34]
mode(s) of income: [0.]
mode(s) of monthly_premium_auto: [65.]
mode(s) of total_claim_amount: [321.6]
median of customer_lifetime_value: 588174.235
median of income: 36234.0
median of monthly_premium_auto: 83.0
median of total_claim_amount: 354.729129
mean of customer_lifetime_value: 793690.2593445694
mean of income: 39295.70121381886
mean of monthly_premium_auto: 193.234360410831
mean of total_claim_amount: 404.98690940896364


#### deviation : varience, stadard deviation, range + coef of variation
All the value are highly variable beacause of the cv>>50% but their is an issu with the monthly_premium_auto because coef of variation, because this value can't be higher than 1. 
Without take in account of this coef, we can see the huge range of eache value in terme of price of policy, customer lifetime value or income

In [114]:
print(f"variance of customer_lifetime_value: {df['customer_lifetime_value'].var(ddof=1) : .2f}")
print(f"variance of income: {df['income'].var(ddof=1) : .2f}")
print(f"variance of monthly_premium_auto: {df['monthly_premium_auto'].var(ddof=1) : .2f}")
print(f"variance of total_claim_amount: {df['total_claim_amount'].var(ddof=1) : .2f}")

variance of customer_lifetime_value:  414064464493.98
variance of income:  928385985.39
variance of monthly_premium_auto:  2563810.60
variance of total_claim_amount:  85864.98


In [115]:
print(f"standard deviation of customer_lifetime_value: {df['customer_lifetime_value'].std() : .2f}")
print(f"standard deviation of income: {df['income'].std() : .2f}")
print(f"standard deviation of monthly_premium_auto: {df['monthly_premium_auto'].std() : .2f}")
print(f"standard deviation of total_claim_amount: {df['total_claim_amount'].std() : .2f}")

standard deviation of customer_lifetime_value:  643478.41
standard deviation of income:  30469.43
standard deviation of monthly_premium_auto:  1601.19
standard deviation of total_claim_amount:  293.03


In [116]:
print(f"max of customer_lifetime_value: {df['customer_lifetime_value'].max()}")
print(f"max of income: {df['income'].max()}")
print(f"max of monthly_premium_auto: {df['monthly_premium_auto'].max()}")
print(f"max of total_claim_amount: {df['total_claim_amount'].max()}")

print(f"min of customer_lifetime_value: {df['customer_lifetime_value'].min()}")
print(f"min of income: {df['income'].min()}")
print(f"min of monthly_premium_auto: {df['monthly_premium_auto'].min()}")
print(f"min of total_claim_amount: {df['total_claim_amount'].min()}")


max of customer_lifetime_value: 5816655.35
max of income: 99960.0
max of monthly_premium_auto: 35354.0
max of total_claim_amount: 2893.239678
min of customer_lifetime_value: 200435.07
min of income: 0.0
min of monthly_premium_auto: 61.0
min of total_claim_amount: 0.382107


In [117]:
clv_range = df['customer_lifetime_value'].max() - df['customer_lifetime_value'].min()
income_range = df['income'].max() - df['income'].min()
monthly_premium_auto_range = df['monthly_premium_auto'].max() - df['monthly_premium_auto'].min()
total_claim_amount_range = df['total_claim_amount'].max() - df['total_claim_amount'].min()
print(f"range of customer_lifetime_value: {clv_range: .2f}")
print(f"range of income: {income_range: .2f}")
print(f"range of monthly_premium_auto: {monthly_premium_auto_range: .2f}")
print(f"range of total_claim_amount: {total_claim_amount_range: .2f}")

range of customer_lifetime_value:  5616220.28
range of income:  99960.00
range of monthly_premium_auto:  35293.00
range of total_claim_amount:  2892.86


In [118]:
cv_clv = df['customer_lifetime_value'].std() / df['customer_lifetime_value'].mean()
cv_income = df['income'].std() / df['income'].mean()
cv_monthly_premium_auto = df['monthly_premium_auto'].std() / df['monthly_premium_auto'].mean()
cv_total_claim_amount = df['total_claim_amount'].std() / df['total_claim_amount'].mean()
print(f"coefficient of variation of customer_lifetime_value: {cv_clv : .2f}")
print(f"coefficient of variation of income: {cv_income : .2f}")
print(f"coefficient of variation of monthly_premium_auto: {cv_monthly_premium_auto : .2f}")
print(f"coefficient of variation of total_claim_amount: {cv_total_claim_amount : .2f}")

coefficient of variation of customer_lifetime_value:  0.81
coefficient of variation of income:  0.78
coefficient of variation of monthly_premium_auto:  8.29
coefficient of variation of total_claim_amount:  0.72


#### position : Quartiles, IRQ and outliers are not required


In [119]:
q1_clv = df['customer_lifetime_value'].quantile(0.25)
q2_clv = df['customer_lifetime_value'].quantile(0.5)
q3_clv = df['customer_lifetime_value'].quantile(0.75)

q1_income = df['income'].quantile(0.25)
q2_income = df['income'].quantile(0.5)
q3_income = df['income'].quantile(0.75)

q1_monthly_premium_auto = df['monthly_premium_auto'].quantile(0.25)
q2_monthly_premium_auto = df['monthly_premium_auto'].quantile(0.5)
q3_monthly_premium_auto = df['monthly_premium_auto'].quantile(0.75)

q1_total_claim_amount = df['total_claim_amount'].quantile(0.25)
q2_total_claim_amount = df['total_claim_amount'].quantile(0.5)
q3_total_claim_amount = df['total_claim_amount'].quantile(0.75)

print(f"q1, q2, q3 of customer_lifetime_value: {q1_clv: .2f}, {q2_clv: .2f}, {q3_clv: .2f}")
print(f"q1, q2, q3 of income: {q1_income: .2f}, {q2_income: .2f}, {q3_income: .2f}")
print(f"q1, q2, q3 of monthly_premium_auto: {q1_monthly_premium_auto: .2f}, {q2_monthly_premium_auto: .2f}, {q3_monthly_premium_auto: .2f}")
print(f"q1, q2, q3 of total_claim_amount: {q1_total_claim_amount: .2f}, {q2_total_claim_amount: .2f}, {q3_total_claim_amount: .2f}")

q1, q2, q3 of customer_lifetime_value:  403407.99,  588174.23,  896287.20
q1, q2, q3 of income:  14072.00,  36234.00,  64631.00
q1, q2, q3 of monthly_premium_auto:  68.00,  83.00,  109.50
q1, q2, q3 of total_claim_amount:  202.16,  354.73,  532.80


Numerical Variables Insights
1. Income (income)
Observations: The mean income is around $39,300, but the median is lower ($36,234). Notably, the mode is 0.

Conclusion: There is a significant segment of customers reporting zero income (which could represent students, unemployed individuals, or missing data). The distribution is quite spread out (CV of 0.78), indicating a diverse customer base ranging from low to high-income earners.

2. Monthly Premium Auto (monthly_premium_auto)
Observations: The mean ($193) is much higher than the median ($83), and the maximum value is an extreme $35,354. This results in a massive Coefficient of Variation (8.29).

Conclusion: This column contains significant outliers or data entry errors. While 75% of customers pay less than $110 per month, the maximum value is unrealistic for a standard auto premium. Data cleaning is required to remove or correct these outliers before further modeling.

3. Total Claim Amount (total_claim_amount)
Observations: The average claim is approximately $405, with values ranging from near zero to $2,893.

Conclusion: This is the most "stable" numerical variable (lowest CV at 0.72). Most claims are concentrated between $202 and $532. The presence of claims near $3,000 suggests that while most accidents are minor, there are occasional high-severity events.

4. Customer Lifetime Value (customer_lifetime_value)
Observations: There is a high standard deviation and a Coefficient of Variation of 0.81.

Conclusion: The "value" of customers to the company varies greatly. This suggests that the company has a mix of long-term, high-value loyalists and newer or lower-tier customers. Segmenting these customers could help in targeted retention strategies.

In [120]:
categorical_cols = df.select_dtypes(include=['object']).columns

for col in categorical_cols:
    print(f"--- Analyse de la colonne : {col} ---")
    print(df[col].value_counts())
    print(f"the main representative value : {df[col].mode().iloc[0] if not df[col].mode().empty else 'No mode'}")
    print("\n")

--- Analyse de la colonne : customer ---
customer
MY31220    1
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
          ..
FL50705    1
WC83389    1
OE15005    1
FV94802    1
BW63560    1
Name: count, Length: 1071, dtype: int64
the main representative value : AA71604


--- Analyse de la colonne : state ---
state
California    331
Oregon        320
Arizona       211
Washington    111
Nevada         98
Name: count, dtype: int64
the main representative value : California


--- Analyse de la colonne : gender ---
gender
Female    502
Male      452
Name: count, dtype: int64
the main representative value : Female


--- Analyse de la colonne : education ---
education
Bachelor                331
College                 313
High School or Below    296
Master                   94
Doctor                   37
Name: count, dtype: int64
the main representative value : Bachelor


--- Analyse de la colonne : customer_lifetime_value ---
customer_lifetime_value
251459.20    4
445811.34    4
578018.2

Categorical Variables Insights
1. State (state)
Observations: California is the most frequent state in the dataset.

Conclusion: The insurance company has its strongest market presence on the West Coast, specifically in California. Marketing efforts should remain focused here while investigating growth opportunities in smaller markets like Washington or Nevada.

2. Gender (gender)
Observations: After cleaning inconsistent values (Femal, female, M, F), the distribution is relatively balanced, with "Female" being the mode.

Conclusion: The customer base is gender-diverse. The insurance products do not show a significant bias toward one specific gender, suggesting a broad market appeal.

3. Education (education)
Observations: "Bachelor" is the most common education level among customers.

Conclusion: The typical customer is highly educated. This profile often correlates with steady income and a preference for clear, detailed policy documentation and professional communication.

4. Policy Type (policy_type)
Observations: "Personal Auto" is the dominant policy type compared to Corporate or Special Auto.

Conclusion: The company’s core business relies on individual private consumers rather than business fleets. The "Personal Auto" segment is the primary driver of policy volume.

5. Vehicle Class (vehicle_class)
Observations: "Four-Door Car" is the most frequent category.

Conclusion: Customers primarily drive standard family or commuter vehicles. Even after grouping luxury categories, the high frequency of four-door cars suggests a customer base focused on utility and everyday transport.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

print(f"mode(s) of customer_lifetime_value: {df['customer_lifetime_value'].mode().values}")
print(f"mode(s) of income: {df['income'].mode().values}")
print(f"mode(s) of monthly_premium_auto: {df['monthly_premium_auto'].mode().values}")
print(f"mode(s) of total_claim_amount: {df['total_claim_amount'].mode().values}")

In [126]:
df['state'] = df['state'].fillna("Unknown")
df['gender'] = df['gender'].fillna("Undefined")



In [127]:
# value_counts() calcule la fréquence de chaque état
location_series = df['state'].value_counts()
bottom_5_locations = location_series.sort_values(ascending=True).head(5)
print("Top 5 less common customer locations:")
print(bottom_5_locations)

Top 5 less common customer locations:
state
Nevada         98
Washington    111
Arizona       211
Oregon        320
California    331
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [131]:
# creer une serie panda
total_sold_policies = df['policy_type'].value_counts()
print("Total sold policies by type:")
print(total_sold_policies)

Total sold policies by type:
policy_type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [141]:
# 1. Création des deux dataframes distincts avec .loc
df_personal = df.loc[df['policy_type'] == 'Personal Auto']
df_corporate = df.loc[df['policy_type'] == 'Corporate Auto']

# 2. Calcul de la moyenne pour la colonne 'income'
# Note : On utilise .mean() et non .avg()
avg_personal = df_personal['income'].mean()
avg_corporate = df_corporate['income'].mean()
print(f"Average income for Personal Auto policy holders: {avg_personal:.2f}")
print(f"Average income for Corporate Auto policy holders: {avg_corporate:.2f}")

if avg_personal > avg_corporate:
    print("Personal Auto policy holders have a higher average income.")
else:
    print("Corporate Auto policy holders have a higher average income.")



Average income for Personal Auto policy holders: 38180.70
Average income for Corporate Auto policy holders: 41390.31
Corporate Auto policy holders have a higher average income.


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [125]:
# Your code here